generator/templates/mock_webnn.h

```diff
@@ -0,0 +1,140 @@
+//* Copyright 2017 The Dawn Authors
+//*
+//* Licensed under the Apache License, Version 2.0 (the "License");
+//* you may not use this file except in compliance with the License.
+//* You may obtain a copy of the License at
+//*
+//*     http://www.apache.org/licenses/LICENSE-2.0
+//*
+//* Unless required by applicable law or agreed to in writing, software
+//* distributed under the License is distributed on an "AS IS" BASIS,
+//* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//* See the License for the specific language governing permissions and
+//* limitations under the License.
+
+#ifndef MOCK_WEBNN_H
+#define MOCK_WEBNN_H
+
+#include <webnn/webnn_proc_table.h>
+#include <webnn/webnn.h>
+#include <gmock/gmock.h>
+
+#include <memory>
+
+// An abstract base class representing a proc table so that API calls can be mocked. Most API calls
+// are directly represented by a delete virtual method but others need minimal state tracking to be
+// useful as mocks.
+class ProcTableAsClass {
+    public:
+        virtual ~ProcTableAsClass();
+
+        void GetProcTableAndDevice(WebnnProcTable* table);
+
+        // Creates an object that can be returned by a mocked call as in WillOnce(Return(foo)).
+        // It returns an object of the write type that isn't equal to any previously returned object.
+        // Otherwise some mock expectation could be triggered by two different objects having the same
+        // value.
+        {% for type in by_category["object"] %}
+            {{as_cType(type.name)}} GetNew{{type.name.CamelCase()}}();
+        {% endfor %}
+
+        {% for type in by_category["object"] %}
+            {% for method in type.methods if len(method.arguments) < 10 and not has_callback_arguments(method) %}
+                virtual {{as_cType(method.return_type.name)}} {{as_MethodSuffix(type.name, method.name)}}(
+                    {{-as_cType(type.name)}} {{as_varName(type.name)}}
+                    {%- for arg in method.arguments -%}
+                        , {{as_annotated_cType(arg)}}
+                    {%- endfor -%}
+                ) = 0;
+            {% endfor %}
+            virtual void {{as_MethodSuffix(type.name, Name("reference"))}}({{as_cType(type.name)}} self) = 0;
+            virtual void {{as_MethodSuffix(type.name, Name("release"))}}({{as_cType(type.name)}} self) = 0;
+        {% endfor %}
+
+	// Special cased mockable methods
+        virtual void OnCompilationComputeCallback(WebnnCompilation self,
```

Comment:
In Dawn we now autogenerate these.

Reply:
Great, fixed in the new commit https://github.com/webmachinelearning/webnn-native/pull/1/commits/2240f5dba7c46f33d1a1779234fdcf1d6c6bce78